### DISCUSSION TITLE
How to do real Nested Nodes? Nested Canvas? or maybe Nested Vue-Flow inside Quasar Accordion?

### DISCUSSION THREAD

**Discussion by ghiscoding** [bcakmakoglu/vue-flow#443] [Q&A]
This is related to a previous Discussion #416 which is related to Nested Nodes. As per your previous answer
> That is by design. It's not a trivial change to move the DOM Elements inside nodes. This behavior might change in the future but I can't tell you when and if 😅 I like to keep the changes in Vue Flow somewhat in line with what is happening at React Flow, so users can expect similar usability and behavior regardless if they want to go with React or Vue.

However, we really need this feature and I'm totally newbie on Canvas stuff and searching on Nested Canvas doesn't help too much since I'm new to canvas. What we'd like to do is really a clickable Node that once clicked would open up and reveal (something like a black box that also has start/end handles. So it's quite similar to [Nested Nodes](https://vueflow.dev/examples/nodes/nesting.html) but we're looking for real nested nodes (not just to drag but also nested in the DOM). Can we use Vue-Flow inside a Vue-Flow? If that works, then I could maybe add to that a Quasar According 

I tried to nest a child Vue-Flow inside a Custom Node of a parent Vue-Flow, but it renders nothing and looking at the html it doesn't seem to like it at all. My guess is that the lib wasn't created to support more than 1 Vue-Flow instance at a time, but how else can I really achieve Nested Nodes

```ts
 <q-card>
    <q-card-section horizontal class="text-body1 q-ml-md q-mb-md justify-between">
        ...
    </q-card-section>
    <q-card-section horizontal class="text-body1 q-ml-md q-mb-md justify-between">
        <div style="height: 300px">
            Child Elements
            <VueFlow id="child-node" v-model="childElements"></VueFlow>
         </div>
    </q-card-section>
</q-card>
```
![image](https://user-images.githubusercontent.com/643976/202279699-66321c36-eb27-48d5-a77c-6ed9b01d0e73.png)

What we are trying to achieve is similar to this other lib [Litegraph.js - demo (subgraph)](https://tamats.com/projects/litegraph/editor/), the concept is that when we click on the sub-graph open icon it reveals the inside of that Nested Node (like a black box) and we want to be able to have these kind of Nested Nodes (that can be expanded/collapsed) and that the user could clone easily. 

That's the only missing piece to fully migrate to Vue-Flow 😃 The rest of the lib is just amazing 

![msedge_RXZeFZCHqz](https://user-images.githubusercontent.com/643976/202280431-c521f03e-6f65-44dd-90d0-86a913b775b2.gif)

**Comment by bcakmakoglu** [comment]
Not sure what you mean by nested canvas. Vue Flow isn't using a canvas.

You can use multiple vue flow instances (see the [example](https://vueflow.dev/examples/multi.html)).
You can use as many instances as you want actually, there is no limit, I mean the Documentation Homepage literally has 4 different Flows on the same page side by side.
Nesting isn't much different, you already guessed you had to use a new *id* to force a new state and that's it.

The code you're showing just means you haven't imported the VueFlow component in your node, hence the tag isn't rendered by Vue. Check you console warnings.

**Reply by ghiscoding** [reply]
oh you're right again, I just needed to `import { VueFlow } from '@vue-flow/core'` to get it working. As for the demo you referenced, it does show 2 Flows but they're side by side, however in our case it would have to be nested inside (something like below). I can probably extend on what I just got working (thanks again to you), it's a good start and might be enough

![image](https://user-images.githubusercontent.com/643976/202282192-cc4c59ff-e74b-46c1-9dce-6033a5593965.png)

**Reply by bcakmakoglu** [reply]
FYI: The parent flows drag handler will probably be on top of your nested flow.
You can add `nodrag` and `nopan` to it, to prevent that. But then you'll notice that your nested flow might possibly be "locked".
You can use a different class name for nodrag and nopan on the nested flow, by setting the props to a different name than the default.
```ts
 noDragClassName?: string
 noWheelClassName?: string
 noPanClassName?: string
```

**Reply by ghiscoding** [reply]
ok thanks for the info, I think I'm already ok because I have the code that you posted on how to resize a node which did disable the drag on the node. However, I didn't know about these extra props, so it's really good to know. Again, thanks a ton for all the info provided ❤️🚀

**Reply by bcakmakoglu** [reply]
Could you check if edges are actually correctly aligned in your nested flow?
When I try to create an example it doesn't seem like the edges are correctly aligned in the nested flow and I want to make sure this behavior isn't just some quirk on my end 😄

**Reply by bcakmakoglu** [reply]
Relevant PR with a Fix for the issue I just described: https://github.com/bcakmakoglu/vue-flow/pull/444

**Reply by ghiscoding** [reply]
this is what it looks like so far and I currently use CSS styling for positioning the handles, the edges seems ok on my side I think
![image](https://user-images.githubusercontent.com/643976/202297504-7dbafa00-ebb9-4181-9cbb-7d0b1769e28f.png)

**Reply by bcakmakoglu** [reply]
That looks fine - strange why my very basic flows have misaligned edges.
The fix should be okay anyway; I'll add it to the next release. If it ends up being an issue, let me know and I'll revert it if necessary 😄

**Reply by ghiscoding** [reply]
~oh wait I do see a problem when trying to connect the edge from a node to the other, it's quite off indeed (see at the end of this animated gif)~ it seems fine after reloading the page, just might be a Hot Reload related issue, so we can disregard it.

**Reply by ghiscoding** [reply]
another problem that I have now is that it's not pannable in the nested Vue-Flow. Am I missing something to get the panning work in the nested Vue-Flow as well? I can use pan in the parent just not in the nested one

```html
<div style="height: 250px; width: 350px">
    Child Elements
    <VueFlow
        id="child-node"
        v-model="childElements"
        fit-view-on-init
        :default-edge-options="{ type: 'custom' }"
        @edge-update="onEdgeUpdate"
        @connect="onConnect"
    >
        <template #edge-custom="props">
            <custom-edge v-bind="props" />
        </template>
        <Background :variant="BackgroundVariant.Lines" />
    </VueFlow>
</div>
```
![msedge_GyCPGsClpW](https://user-images.githubusercontent.com/643976/202302106-4e4aa343-9bf4-4703-be23-844b2372199c.gif)

**Reply by bcakmakoglu** [reply]
Mh... yeah that might be more troublesome than anticipated, the transform of the parent is kind of getting in the way as handle calculation relies on `getBoundingClientRect` which needs to take transforms into account 😞 
The proposed solution isn't cutting it either :/

The issue is that the zoom and pan of the parent container modifies the visible position of the nodes, so when you trigger a connection line, the handle "thinks" it's somewhere different than it visibly is.
As long as the parent has a zoom level that is anything besides 1 it's not gonna work.

I guess it'd be easier to replicate what you showed in the first screenshot like this: Click a button, open a completely new flow outside the current one (for example inside a Modal that is *not* nested inside Vue Flow) and close it when done.

Until I can find a viable solution to this, that's probably the only good way to go about this, sorry.

**Reply by ghiscoding** [reply]
ok I can maybe keep the code I have for the user to see the nodes and if he needs to edit it then I could open a Modal like you said. They would probably prefer without the Modal but at least there's a Plan B, so better than nothing hehe. Let me know if you do find a solution though, thanks 😉 

EDIT
Actually, I think that I'm not able to pan because the parent node had the "nodrag" css class that probably blocks the panning. However, if I remove the "nodrag", then it's the entire node that drags which is about the same end result on the nested Vue-Flow, I still can't use panning in it

**Reply by bcakmakoglu** [reply]
> FYI: The parent flows drag handler will probably be on top of your nested flow. You can add `nodrag` and `nopan` to it, to prevent that. But then you'll notice that your nested flow might possibly be "locked". You can use a different class name for nodrag and nopan on the nested flow, by setting the props to a different name than the default.
> 
> ```ts
>  noDragClassName?: string
>  noWheelClassName?: string
>  noPanClassName?: string
> ```

Have you tried setting a different nodrag/nopan class name on the nested flow? You need to add nodrag / nopan to the nested flow otherwise you can't start interacting with it.

```vue
<template>
  <div class="nopan nodrag" style="background: white; border: 1px solid black; height: 100vh; width: 100vw">
    <VueFlow id="nested" v-model="elements" no-pan-class-name="nopan2" no-drag-class-name="nodrag2" />
  </div>
</template>

**Reply by ghiscoding** [reply]
wow ok it fixes the padding.... but it brings another problem, now I can't connect handles anymore, everything starts to be draggable

Also, I still don't understand how these `no-x` works though, "nopan2" and "nodrag2" are not anywhere in my code but it only works with them defined (even if they're empty!?). I can find the interface in the docs, but no explanation of them, perhaps I could look at React-Flow

Here's my working code (the div `class="nopan nodrag"` has no effect)
```html
<div style="background: white; height: 250px; width: 350px">
    <VueFlow
        id="child-node"
        v-model="childElements"
        fit-view-on-init
        :default-edge-options="{ type: 'custom' }"
        @edge-update="onEdgeUpdate"
        @connect="onConnect"
        no-pan-class-name="nopan2"
        no-drag-class-name="nodrag2"
    >
        <template #edge-custom="props">
            <custom-edge v-bind="props" />
        </template>
        <Background :variant="BackgroundVariant.Lines" />
    </VueFlow>
</div>
```

I tried to cancel the node drag event but that doesn't help even if I do see it go inside my condition. 
```ts
function handleDrag(e: any) {
    if ((e.event.target as HTMLElement).classList.contains('vue-flow__handle')) {
        e.event.preventDefault();
        e.event.stopImmediatePropagation();
    }
}

<VueFlow
    id="child-node"
    ...
    @node-drag-start="handleDrag"
    @node-drag="handleDrag"
    @node-drag-stop="handleDrag"
    no-pan-class-name="nopan2"
    no-drag-class-name="nodrag2"
>
</VueFlow>
```

EDIT 

oh wait it looks like I don't need the `no-drag-class-name="nodrag2"` and then it all works and now I do see the previous issue about misaligned edges (hopefully your last PR fixes that). 

Note that the Edge misaligned is because I played with zoom in/out, when I'm loading the page and not play with the zoom, then the edge is not misaligned. 

![msedge_PGaruzzrml](https://user-images.githubusercontent.com/643976/202317164-ca161351-0deb-4f46-984a-7f860ed954da.gif)

**Reply by ghiscoding** [reply]
@bcakmakoglu I'm not sure if you read my last comment since I edited it couple times, but will you have a fix for the misaligned Edges (shown in animated gif below and above) in the nested Vue-Flow? I thought PR #444 was going to fix this but I see you closed it, so I'm unsure if you have a possible fix for this? It seems to happen when I zoom in/out and pan the parent node, then its nested Flow seems to be confused about its real position

![msedge_DIsZs2zH9N](https://user-images.githubusercontent.com/643976/202506986-2642edd5-401c-4cf0-8bdf-ac865bee6680.gif)

**Reply by bcakmakoglu** [reply]
Yeah I saw that but I gotta disappoint you, I don't think there's going to be a proper fix for that issue any time soon.
As long as the parent applies a transform with a scale, the child will start behaving strangely as you've seen. So no nested Vue Flow components.

You have to place your second Flow outside the transform of the first Flow otherwise you'll always end up having edge misalignments.

**Reply by ghiscoding** [reply]
> You have to place your second Flow outside the transform of the first Flow otherwise you'll always end up having edge misalignments.

do you mean like a Modal that is a completely different component or something else? I would rather keep this nested Vue-Flow approach that I have (if possible), my team likes this POC so far.

**Reply by bcakmakoglu** [reply]
As long as it's outside the transformation pane of your parent, it's good to go.
A Modal would be a good place to start, so something like that

```vue
<VueFlow v-model="elements">
	<Modal v-if="isVisible" :data="data" />
</VueFlow>
```

I can't promise when and *if* that nested flow issue can be solved, at least not for connections.
I can solve the issue for the edge misalignment for edges that already exist, that way you could at least show a static version of that Flow and open an interactive one in a separate context outside the parent flows transformation.

As it's not really a stable implementation, I will add an `experimentalFeatures` flag to opt into this behavior.
Relevant PR: https://github.com/bcakmakoglu/vue-flow/pull/449

(will be released in `1.5.0`)

**Reply by ghiscoding** [reply]
ok great, thanks again for all of that. Will that option be a props on Custom Edge or a props on Vue-Flow? I don't see how to enable it from the PR. Have a nice day

**Reply by bcakmakoglu** [reply]
It's gonna be a prop on Vue Flow, but you should pass it using `useVueFlow` and not component props.

By the way,
[Here's an example](https://codesandbox.io/s/determined-jerry-2uupfi?file=/src/CustomNode.vue) I came up with to solve your current usecase. I guess it's not as cool as you envisioned (I would also prefer if Nested Flows would actually work without misalignments 😄) but it does the trick.

**Reply by ghiscoding** [reply]
great thanks, I got that working, it's definitely good to have multiples POC. Thanks a ton
